### PR TITLE
(moveit_py) Add Trajectory Execution Manager

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
### Description

Add Trajectory Execution Manager.

Note: I have not been able to test the correct behaviour of all added functions. Also I skipped some functions, which are shown as comments.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit2/blob/main/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [x] Include a screenshot if changing a GUI
- [x] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by sjahr** [comment]
@MatthijsBurgh Can you run pre-commit? See here for instructions on how to do that https://moveit.ros.org/documentation/contributing/code/#pre-commit-formatting-checks

**Comment by MatthijsBurgh** [comment]
@sjahr done

**Comment by MatthijsBurgh** [comment]
Sorry for breaking CI again. pre-commit didn't do its jobs correctly. I don't know how.

File: moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+           R"(
+           Add a trajectory for future execution.
+
+           Optionally specify a set of controllers to consider using for the trajectory.
+           Multiple controllers can be used simultaneously to execute the different parts of the trajectory.
+           If multiple controllers can be used, preference is given to the already loaded ones.
+           If no controller is specified, a default is used.
+           )")
+
+      // ToDo(MatthijsBurgh)
```
Comment (henningkayser) [inline moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:157]: would you mind opening an issue with these items and referencing it here?

File: moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+           Set joint-value tolerance for validating trajectory's start point against current robot state.
+           )")
+
+      .def("setWaitForTrajectoryCompletion",
+           &trajectory_execution_manager::TrajectoryExecutionManager::setWaitForTrajectoryCompletion, py::arg("flag"),
+           R"(
+           Enable or disable waiting for trajectory completion.
+           )");
+
+  // ToDo(MatthijsBurgh)
```
Comment (henningkayser) [inline moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:218]: same

File: moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik Inc. nor the names of its
```
Comment (henningkayser) [inline moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.h:17]: the new license headers are still naming PickNik here. You might want to fix that.

**Review by henningkayser — APPROVED** [review]
I gave this a quick pass, the bindings look all correct to me. I would appreciate if @peterdavidfagan or @JafarAbdi could verify and test this.

**Comment by peterdavidfagan** [comment]
> Created an issue and referenced it in the code. Updated the copyright headers of only the changed files in this PR.

🚀🚀🚀

**Comment by MatthijsBurgh** [comment]
@peterdavidfagan friendly ping to merge this PR ;)

**Review by peterdavidfagan — APPROVED** [review]
These changes LGTM. I haven't had the opportunity to test all the added functionalities but I would be in favour of merging these changes and addressing any follow ups in issues.

Thanks @MatthijsBurgh for this contribution 🚀.

File: moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+#include "trajectory_execution_manager.h"
+
+namespace moveit_py
+{
+namespace bind_trajectory_execution_manager
+{
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_py.bind_trajectory_execution_manager");
+
+void init_trajectory_execution_manager(py::module& m)
```
Comment (sjahr) [inline moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:45]:
```suggestion
void initTrajectoryExecutionManager(py::module& m)
```

File: moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+#include <moveit/trajectory_execution_manager/trajectory_execution_manager.h>
+
+namespace py = pybind11;
+
+namespace moveit_py
+{
+namespace bind_trajectory_execution_manager
+{
+
+void init_trajectory_execution_manager(py::module& m);
```
Comment (sjahr) [inline moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.h:52]:
```suggestion
void initTrajectoryExecutionManager(py::module& m);
```

File: moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit_py.bind_trajectory_execution_manager");
+
+void initTrajectoryExecutionManager(py::module& m)
+{
+  py::class_<trajectory_execution_manager::TrajectoryExecutionManager,
+             trajectory_execution_manager::TrajectoryExecutionManagerPtr>(m, "TrajectoryExecutionManager", R"(
+      Manages the trajectory execution.
+      )")
+
+      .def("isManagingControllers", &trajectory_execution_manager::TrajectoryExecutionManager::isManagingControllers,
```
Comment (sjahr) [inline moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:52]:
```suggestion
      .def("is_managing_controllers", &trajectory_execution_manager::TrajectoryExecutionManager::isManagingControllers,
```
Shouldn't the python functions be defined in `snake_case` (`camelCase` is used for cpp functions)? I think it is done like this everywhere else, so we should be consistent about it. Can you update the other function signatures as well?

**Review by sjahr — COMMENTED** [review]
Thanks a lot for this contribution (and your patience regarding reviews). Only one thing left to be changed and then this can finally be merged!

File: moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp (hunk 6 above)
Comment (peterdavidfagan) [reply moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:52]: We need to update many of the existing cpp methods as there are discrepancies elsewhere in the library which may have lead to @MatthijsBurgh using different conventions. 

Not sure if there is an easy way to add this to auto formatter (clang-format).

Comment (peterdavidfagan) [reply moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:52]: Added issue: https://github.com/ros-planning/moveit2/issues/2499

Comment (sjahr) [reply moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:52]: I think that got covered by https://github.com/ros-planning/moveit2/pull/2177 or do you mean something else @peterdavidfagan. That PR also added clang-tidy checks to warn about cpp function names that don't follow the conventions. But don't think there are checks to catch  `.def("...", ..)` errors

Comment (peterdavidfagan) [reply moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:52]: [49052c1](https://github.com/ros-planning/moveit2/pull/2177/commits/49052c11eb0ac20e0b7df2c3826721dfadb568ef) ❤️. 

My fork was outdated only seeing this updates now. Thanks so much.

As for `.def` I would personally need to become more familiar with `clang-tidy`, maybe there is a way to get a rule which also flags `.def`. For now I am happy to be more vigilant about this when pushing code or reviewing.

Comment (sjahr) [reply moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:52]: Awesome, thanks for opening the issue though!

Comment (MatthijsBurgh) [reply moveit_py/src/moveit/moveit_ros/trajectory_execution_manager/trajectory_execution_manager.cpp:52]: So the python functions will be snake_case?

**Comment by MatthijsBurgh** [comment]
@sjahr I fixed the styling. I think then this one is ready to be merged too, correct?

**Review by sjahr — APPROVED** [review]

**Pull request merged**